### PR TITLE
[1/N] Auto-generate data sources documentation

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Generate data docs from a jinja template. Not complete, will come back and add docs for each data source later.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -86,6 +88,57 @@ def _transform_changelog(path_in: str, path_out: str) -> None:
 with open(os.path.join(generated_dir, "providers.rst"), "w") as f:
     f.writelines(sorted(lines, key=str.casefold))
 
+# -- Generate data docs ------------------------------------------------------
+
+generated_dir = os.path.join("reference", "generated")
```
Comment (Borda) [inline docs/source/conf.py:93]: let's move it to a special script... :]

**Review by Borda — COMMENTED** [review]
not sure seems to me like trying to substitute documentation, I would rather include it directly in the docstring of the tsk

**Review by ananyahjha93 — APPROVED** [review]

File: .gitignore [hunk 2]
```diff
@@ -79,6 +79,7 @@ docs/api/
 docs/notebooks/
 docs/source/api/generated/
 docs/source/integrations/generated/
+docs/source/reference/generated/
```
Comment (Borda) [inline .gitignore:82]: is this still valid?

**Review by Borda — COMMENTED** [review]
need to see it live :]
https://lightning-flash.readthedocs.io/en/feature-auto_data_docs

**Review by akihironitta — COMMENTED** [review]
I like this idea! You might be aware, but we need to either remove a duplicated page or exclude one from search results:
- https://2705-333857397-gh.circle-artifacts.com/0/html/reference/generated/ImageClassificationData.html
- https://2705-333857397-gh.circle-artifacts.com/0/html/reference/image_classification.html#loading-data

**Comment by ethanwharris** [comment]
@akihironitta Switched for a sphinx extension, so now no generated pages!

**Review by tchaton — APPROVED** [review]
Looks great !

**Review by akihironitta — APPROVED** [review]
Looks good to me!

**Pull request merged**